### ISSUE TITLE
Cannot get `no-floating-promises` to work

### ISSUE THREAD

**Issue by dscho** [palantir/tslint#3929]
### Bug Report

- __TSLint version__: 5.10.0
- __TypeScript version__: Version 2.8.3
- __Running TSLint via__: CLI (`node_modules\.bin\tslint --project tsconfig.json *.ts *\*.ts`)

#### TypeScript code being linted

```ts
import * as fs from "fs";
import * as util from "util";
const stat = util.promisify(fs.stat);

export async function isDirectory(path: string): Promise<boolean> {
    try {
        if ((await stat(path)).isDirectory()) {
            return true;
        }
    } catch (reason) {
        /* it's okay */
    }
    return false;
}

(async (): Promise<void> => {
    if (isDirectory("123")) {
        console.log("well, hello 123!");
    }
})();
```

with `tslint.json` configuration:

```json
{
    "extends": "tslint:recommended",
    "rules": {
        "max-line-length": {
            "options": [80]
        },
        "no-floating-promises": true,
        "no-unused-expression": true,
        "no-unused-variable": true
    }
}
```

#### Actual behavior

TSLint does not warn about the return value of `isDirectory()` (called in the 4th-last line in the code snippet) being a `Promise` that is never handled (missing `await`).

#### Expected behavior

I would hope that TSLint saves me from the many `await`s I regularly omit by mistake.

**Comment by JKillian** [comment]
Looks like TSLint is examining expressions inside of conditionals; seems like a bug that should get fixed!

**Comment by dscho** [comment]
Oh, and it would appear that my understanding of what `no-floating-promises` is incorrect :-( There is [an explicit test in `test/rules/no-floating-promises/promises/test.ts.lint` to verify that it does *not* report an unawaited Promise in a conditional](https://github.com/palantir/tslint/blob/4a11f952641bd6b686644fda2117786a13283650/test/rules/no-floating-promises/promises/test.ts.lint#L121):

```typescript
while (returnsPromiseFunction());
```

So now I need *even more* guidance. Is this a bug in `no-floating-promises`' design? What is the sense of letting `while (returnsPromiseFunction());` go unreported?
